<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: World Development Statistics

--- 
# Part 1

Part 1 requires knowledge of basic Python.

---

## Problem Statement

Decide on your problem statement that will guide your analysis for this project. For guidelines, sample prompts, or inspiration, check out the README.

**To-Do:** *Replace this cell with your problem statement.*

Doctors-Go-Anywhere have reached out after their recent funding, as they are looking to expand their operations. They are wanting to indentify the Top 5 regions globally that have the least Healthcare access and would benefit from their services.

### Contents:
- [Background](#Background)
- [Data Import & Cleaning](#Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-Data)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

## Background

For our first project, we're going to take a look at World Development Statistics from [Gapminder](https://www.gapminder.org/about/), an independent Swedish foundation that aims to make data about the world more accessible and reliable. A good introduction on Gapminder is this [Ted Talk](https://www.ted.com/talks/hans_rosling_the_best_stats_you_ve_ever_seen) from Hans Rosling, which also shows how effective data visualization can be for your audience.

**To-Do:** *Fill out this cell (or edit the above cell) with any other background or information that is necessary for your problem statement.*

Additional will also be pulled from world health organization. We will also be looking into see the number of existing doctors that there are in regions and see how much governments are investing into their healthcare to see how that collerates into in their position on the rankings.


### Choose your Data

There are 3 datasets included in the [`data`](./data/) folder for this project. You are required to pick **at least two** of these to complete your analysis. Feel free to use more than two if you would like, or add other relevant datasets you find online. 

**HINT: There are more datasets available on the Gapminder website**

* [`population.csv`](./data/population.csv): Population by Country
* [`life_expectancy.csv`](./data/life_expectancy.csv): Life Expectancy by Country
* [`gni_per_cap_atlas_method_con2021.csv`](./data/gni_per_cap_atlas_method_con2021.csv): Gross National Income (GNI) per capita in current US dollars

**To-Do:** *Fill out this cell with the datasets you will use for your analysis. Write a brief description of the contents for each dataset that you choose.*

Alongside the population & life expectancy data files, I will be pulling data on the number of doctors, health aid given and total health spending.


### Outside Research

Based on your problem statement and your chosen datasets, spend some time doing outside research on state policies or additional information that might be relevant. Summarize your findings below. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. **Make sure that you cite your sources.**

**To-Do:** *Fill out this cell with outside research or any additional background information that will support your analysis.*

### Coding Challenges

1. Manually calculate mean:

    Write a function that takes in values and returns the mean of the values. Create a list of numbers that you test on your function to check to make sure your function works!
    
    *Note*: Do not use any mean methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [1]:
# Code:
list_numbers = [56, 577, 343, 332, 566, 21, 96]
mean = sum(list_numbers)/len(list_numbers); 

2. Manually calculate standard deviation:

    The formula for standard deviation is below:

    $$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

    Where $x_i$ represents each value in the dataset, $\mu$ represents the mean of all values in the dataset and $n$ represents the number of values in the dataset.

    Write a function that takes in values and returns the standard deviation of the values using the formula above. Hint: use the function you wrote above to calculate the mean! Use the list of numbers you created above to test on your function.
    
    *Note*: Do not use any standard deviation methods built-in to any Python libraries to do this! This should be done without importing any additional libraries.

In [2]:
# Code:
((sum((num - mean)**2 for num in list_numbers))/len(list_numbers)) ** 0.5


216.35081111331144

3. Data cleaning function:
    
    Write a function that takes in a string consisting a number and an abbreviated place value (ex. '5M', '45K'), and converts that string to an integer. For example, inputting '5M' in your function should return 5000000, '45K' should return 45000, etc. Make sure to test your function to make sure it works!

You will use these functions later on in the project!

In [78]:
# Code:
def string_tonum(org_fig):
    keys = {'K': 1000, 'M': 1000000, 'B': 1000000000, '0': 1}

    num = float(org_fig[:-1])
    letter = org_fig[-1].upper()
    
    if letter in keys:
        return num * keys[letter]
    
print(string_tonum('5M'))

5000000.0


In [79]:
popul23[popul23['2005'].astype(str).apply(lambda s: s[-1]).isin(['n', '0'])]

country  2000  2001  2002  2003  2004  2005 2006   2007   2008  ...  \
74   Holy See   779   NaN   NaN   NaN   NaN   NaN  NaN    NaN    NaN  ...   
180    Tuvalu  9640  9620  9610  9670  9790  9910  10k  10.1k  10.3k  ...   

      2014   2015   2016   2017   2018 2019   2020   2021   2022   2023  
74     NaN    NaN    NaN    NaN    NaN  NaN    NaN    NaN    NaN    NaN  
180  10.9k  10.9k  10.9k  10.8k  10.9k  11k  11.1k  11.2k  11.3k  11.4k  

[2 rows x 25 columns]

--- 
# Part 2

Part 2 requires knowledge of Pandas, EDA, data cleaning, and data visualization.

---

*All libraries used should be added here*

In [4]:
# Imports:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [88]:
#Data files being used

population = pd.read_csv('../data/population.csv')
life_expect = pd.read_csv('../data/life_expectancy.csv')
health_aid = pd.read_csv('../data/health_aid_given_percent_of_aid.csv')
medical_doc = pd.read_csv('../data/medical_doctors_per_1000_people.csv')

## Data Import and Cleaning

### Data Import & Cleaning

Import the datasets that you selected for this project and go through the following steps at a minimum. You are welcome to do further cleaning as you feel necessary:
1. Display the data: print the first 5 rows of each dataframe to your Jupyter notebook.
2. Check for missing values.
3. Check for any obvious issues with the observations (keep in mind reasonable values for each column).
4. Fix any errors you identified in steps 2-3.
5. Display the data types of each feature.
6. Fix any incorrect data types found in step 5.
    - Fix any individual values preventing other columns from being the appropriate type.
    - If your dataset has a column of abbreviated numbers (ex. '3M', '45K', etc.), use the function you wrote in Part 1 (coding challenges, number 3) to convert this to integers! *Hint*: use `.map()` or `.apply()`.
7. Rename Columns.
    - Column names should be all lowercase.
    - Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`).
    - Column names should be unique and informative.
8. Drop unnecessary rows (if needed).
9. Merge dataframes that can be merged.
10. Perform any additional cleaning that you feel is necessary.
11. Save your cleaned and merged dataframes as csv files.

In [80]:
# Code population cleaning
population.head()
popul23 = population[['country'] + [str(year) for year in range(2000, 2024)]]
missing_pop = popul23[popul23.isnull().any(axis=1)]
print(missing_pop)



Empty DataFrame
Columns: [country, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023]
Index: []

[0 rows x 25 columns]


In [76]:
#removing Holy See from dataset as too many dates points missing
population = population.dropna() #dropping Holy See for a lot of missing data
popul23.head()

country   2000   2001   2002   2003   2004   2005   2006  \
0           Afghanistan  19.5M  19.7M    21M  22.6M  23.6M  24.4M  25.4M   
1                Angola  16.4M  16.9M  17.5M  18.1M  18.8M  19.5M  20.2M   
2               Albania  3.18M  3.15M  3.12M  3.09M  3.06M  3.03M     3M   
3               Andorra  66.1k  67.8k  70.8k  73.9k  76.9k  79.8k  80.2k   
4  United Arab Emirates  3.28M  3.45M  3.63M  3.81M  3.99M  4.28M   4.9M   

    2007   2008  ...   2014   2015   2016   2017   2018   2019   2020   2021  \
0  25.9M  26.4M  ...  32.7M  33.8M  34.6M  35.6M  36.7M  37.8M    39M  40.1M   
1  20.9M  21.7M  ...  27.1M  28.1M  29.2M  30.2M  31.3M  32.4M  33.4M  34.5M   
2  2.98M  2.95M  ...  2.88M  2.88M  2.88M  2.88M  2.88M  2.87M  2.87M  2.85M   
3  78.2k  76.1k  ...  71.6k  71.7k  72.5k  73.8k    75k  76.3k  77.7k    79k   
4  5.87M  6.99M  ...  8.84M  8.92M  8.99M  9.07M  9.14M  9.21M  9.29M  9.37M   

    2022   2023  
0  41.1M  42.2M  
1  35.6M  36.7M  
2  2.84M  2.83M  
3  79.8k  80.1k  
4  9.44M  9.52M  

[5 rows x 25 columns]

In [81]:
popul23.dtypes

country    object
2000       object
2001       object
2002       object
2003       object
2004       object
2005       object
2006       object
2007       object
2008       object
2009       object
2010       object
2011       object
2012       object
2013       object
2014       object
2015       object
2016       object
2017       object
2018       object
2019       object
2020       object
2021       object
2022       object
2023       object
dtype: object

In [85]:
#population = population.drop(74) #74 for the row that Holy see was on
#running missing statement again to ensure it has been removed
#popul23.isnull().sum()

#Converting the numbers from abbreviations
for yrs in popul23.columns[1:]:
    popul23[yrs] = popul23[yrs].map(string_tonum)

#popul23.iloc[:, 1:] = popul23.iloc[:, 1:].apply(string_tonum)   

/var/folders/pj/jq7g3ftn5557342n7bvxr3q80000gn/T/ipykernel_72096/329833947.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  popul23[yrs] = popul23[yrs].map(string_tonum)


In [86]:
popul23

country        2000        2001        2002        2003  \
0             Afghanistan  19500000.0  19700000.0  21000000.0  22600000.0   
1                  Angola  16400000.0  16900000.0  17500000.0  18100000.0   
2                 Albania   3180000.0   3150000.0   3120000.0   3090000.0   
3                 Andorra     66100.0     67800.0     70800.0     73900.0   
4    United Arab Emirates   3280000.0   3450000.0   3630000.0   3810000.0   
..                    ...         ...         ...         ...         ...   
192                 Samoa    184000.0    186000.0    187000.0    187000.0   
193                 Yemen  18600000.0  19100000.0  19700000.0  20200000.0   
194          South Africa  46800000.0  47200000.0  47700000.0  48100000.0   
195                Zambia   9890000.0  10200000.0  10500000.0  10800000.0   
196              Zimbabwe  11800000.0  11900000.0  12000000.0  12100000.0   

           2004        2005        2006        2007        2008  ...  \
0    23600000.0  24400000.0  25400000.0  25900000.0  26400000.0  ...   
1    18800000.0  19500000.0  20200000.0  20900000.0  21700000.0  ...   
2     3060000.0   3030000.0   3000000.0   2980000.0   2950000.0  ...   
3       76900.0     79800.0     80200.0     78200.0     76100.0  ...   
4     3990000.0   4280000.0   4900000.0   5870000.0   6990000.0  ...   
..          ...         ...         ...         ...         ...  ...   
192    188000.0    189000.0    189000.0    190000.0    192000.0  ...   
193  20700000.0  21300000.0  22000000.0  22600000.0  23300000.0  ...   
194  48600000.0  49000000.0  49500000.0  50000000.0  50600000.0  ...   
195  11200000.0  11600000.0  12000000.0  12400000.0  12900000.0  ...   
196  12200000.0  12200000.0  12300000.0  12500000.0  12600000.0  ...   

           2014        2015        2016        2017        2018        2019  \
0    32700000.0  33800000.0  34600000.0  35600000.0  36700000.0  37800000.0   
1    27100000.0  28100000.0  29200000.0  30200000.0  31300000.0  32400000.0   
2     2880000.0   2880000.0   2880000.0   2880000.0   2880000.0   2870000.0   
3       71600.0     71700.0     72500.0     73800.0     75000.0     76300.0   
4     8840000.0   8920000.0   8990000.0   9070000.0   9140000.0   9210000.0   
..          ...         ...         ...         ...         ...         ...   
192    202000.0    204000.0    206000.0    208000.0    210000.0    212000.0   
193  27800000.0  28500000.0  29300000.0  30000000.0  30800000.0  31500000.0   
194  54700000.0  55900000.0  56400000.0  56600000.0  57300000.0  58100000.0   
195  15700000.0  16200000.0  16800000.0  17300000.0  17800000.0  18400000.0   
196  13900000.0  14200000.0  14500000.0  14800000.0  15100000.0  15400000.0   

           2020        2021        2022        2023  
0    39000000.0  40100000.0  41100000.0  42200000.0  
1    33400000.0  34500000.0  35600000.0  36700000.0  
2     2870000.0   2850000.0   2840000.0   2830000.0  
3       77700.0     79000.0     79800.0     80100.0  
4     9290000.0   9370000.0   9440000.0   9520000.0  
..          ...         ...         ...         ...  
192    215000.0    219000.0    222000.0    226000.0  
193  32300000.0  33000000.0  33700000.0  34400000.0  
194  58800000.0  59400000.0  59900000.0  60400000.0  
195  18900000.0  19500000.0  20000000.0  20600000.0  
196  15700000.0  16000000.0  16300000.0  16700000.0  

[196 rows x 25 columns]

In [87]:
#Saving the cleaned population file
popul23.to_csv('Population23.csv', index = False)

In [93]:
#Life Expectancy cleaning
life_expect.head()


country  1800  1801  1802  1803  1804  1805  1806  1807  1808  \
0           Afghanistan  28.2  28.2  28.2  28.2  28.2  28.2  28.1  28.1  28.1   
1                Angola  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0  27.0   
2               Albania  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4  35.4   
3               Andorra   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  United Arab Emirates  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7  30.7   

   ...  2091  2092  2093  2094  2095  2096  2097  2098  2099  2100  
0  ...  75.5  75.7  75.8  76.0  76.1  76.2  76.4  76.5  76.6  76.8  
1  ...  78.8  79.0  79.1  79.2  79.3  79.5  79.6  79.7  79.9  80.0  
2  ...  87.4  87.5  87.6  87.7  87.8  87.9  88.0  88.2  88.3  88.4  
3  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4  ...  82.4  82.5  82.6  82.7  82.8  82.9  83.0  83.1  83.2  83.3  

[5 rows x 302 columns]

In [104]:
#creating new variable with only 2000 to 2023
life_expect23 = life_expect[['country'] + [str(year) for year in range(2000, 2024)]]

In [105]:
#locating missing data
life_expect23[life_expect23.isnull().any(axis=1)]

country  2000  2001  2002  2003  2004  2005  2006  2007  \
3                Andorra  80.8  80.9  81.1  81.2  81.3  81.4  81.5  81.7   
46              Dominica  72.1  72.2  72.3  72.2  72.3  72.6  72.6  72.5   
93   St. Kitts and Nevis  69.9  70.3  70.7  71.1  71.2  71.4  71.5  71.7   
107               Monaco  78.8  78.9  78.9  79.0  79.0  79.0  79.0  79.1   
112     Marshall Islands  62.7  63.0  63.0  63.0  63.1  63.1  63.2  63.4   
131                Nauru  59.1  58.9  58.8  58.7  58.7  58.8  58.9  59.0   
138                Palau  66.4  66.5  66.7  66.8  66.8  66.8  66.9  67.0   
156           San Marino  81.5  81.6  81.7  81.7  81.6  81.6  81.6  81.7   
178               Tuvalu  61.5  64.7  65.1  65.3  65.5  65.7  66.0  66.1   

     2008  ...  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023  
3    81.8  ...  82.0  82.0  82.1  82.1  82.1  82.2   NaN   NaN   NaN   NaN  
46   72.7  ...  72.2  71.5  72.3  71.2  72.3  72.3   NaN   NaN   NaN   NaN  
93   72.0  ...  72.1  72.0  72.1  72.1  72.1  72.1   NaN   NaN   NaN   NaN  
107  79.1  ...  79.4  79.5  79.7  79.8  79.9  80.1   NaN   NaN   NaN   NaN  
112  63.6  ...  64.8  64.9  65.1  65.2  65.3  65.5   NaN   NaN   NaN   NaN  
131  59.3  ...  61.9  62.3  62.6  63.0  63.2  63.4   NaN   NaN   NaN   NaN  
138  67.0  ...  67.1  67.2  67.3  67.4  67.4  67.5   NaN   NaN   NaN   NaN  
156  81.8  ...  82.0  82.0  82.0  82.0  82.1  82.2   NaN   NaN   NaN   NaN  
178  66.3  ...  67.2  67.4  67.5  67.7  67.8  67.9   NaN   NaN   NaN   NaN  

[9 rows x 25 columns]

In [106]:
#Due to the missing data above dropping years 2020-2023
life_expect19 = life_expect23[['country'] + [str(year) for year in range(2000, 2020)]]
life_expect19.head()

country  2000  2001  2002  2003  2004  2005  2006  2007  2008  \
0           Afghanistan  54.7  54.8  55.5  56.5  57.1  57.6  58.0  58.5  59.2   
1                Angola  52.8  53.4  54.5  55.1  55.5  56.4  57.0  58.0  58.8   
2               Albania  75.4  76.0  75.9  75.6  75.8  76.2  76.9  77.5  77.6   
3               Andorra  80.8  80.9  81.1  81.2  81.3  81.4  81.5  81.7  81.8   
4  United Arab Emirates  69.1  69.2  69.4  69.3  69.1  69.2  69.5  70.0  70.4   

   ...  2010  2011  2012  2013  2014  2015  2016  2017  2018  2019  
0  ...  60.5  61.0  61.4  61.9  61.9  61.9  62.0  62.9  62.7  63.3  
1  ...  60.2  60.8  61.4  62.1  63.0  63.5  63.9  64.2  64.6  65.1  
2  ...  78.1  78.1  78.2  78.3  78.2  78.1  78.2  78.3  78.4  78.5  
3  ...  81.8  81.9  81.9  82.0  82.0  82.0  82.1  82.1  82.1  82.2  
4  ...  70.8  71.0  71.2  71.6  73.0  73.2  73.4  73.5  73.7  73.9  

[5 rows x 21 columns]

In [107]:
#Health aid data cleaning
health_aid.head()

country   1971  1972   1973    1974   1975    1976     1977     1978  \
0    Australia  0.945  1.20   5.42   4.930  10.90   9.390   0.8260   3.5100   
1      Austria    NaN   NaN   1.65   0.136   9.68   0.633   0.0383   0.0384   
2      Belgium  2.930  2.89  13.40  19.900  15.80  11.300  15.9000  17.1000   
3       Canada  0.838  1.16   2.15   2.530   3.14   1.970   1.9900   2.5800   
4  Switzerland  1.650  1.45   7.61   3.780   0.12   0.996   5.5600   4.6800   

     1979  ...   1998   1999   2000   2001   2002   2003   2004   2005   2006  \
0   5.340  ...   6.32   9.78  11.00  12.30  12.30  10.90  10.00   8.71  11.10   
1   0.118  ...   4.22  24.60  14.80   2.56   4.30   6.59  11.10  11.40   6.74   
2  16.300  ...  20.70  20.60  18.30  16.10  12.80  14.20  11.50  13.20  12.60   
3   0.831  ...   5.73   3.83   3.78   8.77   7.97  13.10   7.89  18.60  13.00   
4   8.480  ...   5.10   9.37   7.97   8.91   7.05   4.95   5.61   5.38   6.25   

    2007  
0   8.55  
1   7.69  
2  13.90  
3  22.50  
4   6.90  

[5 rows x 38 columns]

In [110]:
health_aid23 = health_aid[['country'] + [str(year) for year in range(2000,2008)]]
health_aid23

country   2000   2001   2002   2003   2004   2005   2006   2007
0     Australia  11.00  12.30  12.30  10.90  10.00   8.71  11.10   8.55
1       Austria  14.80   2.56   4.30   6.59  11.10  11.40   6.74   7.69
2       Belgium  18.30  16.10  12.80  14.20  11.50  13.20  12.60  13.90
3        Canada   3.78   8.77   7.97  13.10   7.89  18.60  13.00  22.50
4   Switzerland   7.97   8.91   7.05   4.95   5.61   5.38   6.25   6.90
5       Germany   2.11   2.69   5.50   3.82   2.67   2.70   4.36   4.19
6       Denmark   3.38   7.13   8.76  12.00  10.80   7.25  14.20   8.19
7         Spain   8.53  11.00   9.66   9.38   9.43  10.30   7.46   7.40
8       Finland   7.56   6.92  15.70  11.10   5.87   5.76  12.20   5.43
9        France   5.24   7.31   8.50   6.20   8.16   7.55   5.74   3.31
10           UK  15.40  10.80  17.00  13.90   8.30   9.48  12.30  12.00
11       Greece   5.44   7.12   3.39  10.40  16.80  21.60  17.00  13.60
12      Ireland  21.50  20.90  32.10  30.20  31.10  31.50  18.00  23.40
13        Italy  11.80   5.09   8.90  20.60   8.61  14.00  13.00  14.30
14        Japan   3.04   3.57   4.30   4.02   6.66   2.22   3.70   3.37
15   Luxembourg  23.80    NaN    NaN    NaN  26.10  26.00  23.60  21.20
16  Netherlands   8.49   8.08   8.68   8.68   7.58   5.21  12.80   4.44
17       Norway   4.38  14.80   8.18  10.80   7.54  11.50  12.30   7.28
18  New Zealand   5.65   5.96   5.54   9.26   8.32  11.50   9.24   5.96
19     Portugal   4.46   5.25   4.90   5.00   7.04   5.55   5.61   4.25
20       Sweden  12.40   3.58   8.20   8.61   9.27   8.89  10.70  11.80
21          USA   5.32   7.38   8.84   5.88   5.66   6.59   8.00   6.10

In [ ]:


### Data Dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

*Note*: if you are unsure of what a feature is, check the source of the data! This can be found in the README.

**To-Do:** *Edit the table below to create your own data dictionary for the datasets you chose.*

|Feature|Type|Dataset|Description|
|---|---|---|---|
|column name|int/float/object|This is an example| 


## Exploratory Data Analysis

Complete the following steps to explore your data. You are welcome to do more EDA than the steps outlined here as you feel necessary:
1. Summary Statistics.
2. Use a **dictionary comprehension** to apply the standard deviation function you create in part 1 to each numeric column in one of your dataframes.  **No loops**.
    - Assign the output to variable `sd` as a dictionary where: 
        - Each column name is now a key 
        - That standard deviation of the column is the value
        - *Example Output :* `{'1801': 3.80, '1802': 3.93, ...}`
3. Investigate trends in the data.
    - Using sorting and/or masking (along with the `.head()` method to avoid printing our entire dataframe), consider questions relevant to your problem statement. Some examples are provided below (but feel free to change these questions for your specific problem):
        - Which states have the highest and lowest populations for the past few decades?
        - Which states have the highest and lowest economic growth over the course of the dataset?
        - What years were there new countries created?
        - Do any countries have higher growth rates of population than their economy?
        - Which countries have had the lowest population growth?
    - **You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [ ]:
#Code:

**To-Do:** *Edit this cell with your findings on trends in the data (step 3 above).*

## Visualize the Data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers. It is important to not only create visualizations, but to **interpret your visualizations** as well.

**Every plot should**:
- Have a title
- Have axis labels
- Have appropriate tick labels
- Text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Have an interpretation to aid understanding

Here is an example of what your plots should look like following the above guidelines. Note that while the content of this example is unrelated, the principles of visualization hold:

![](https://snag.gy/hCBR1U.jpg)
*Interpretation: The above image shows that as we increase our spending on advertising, our sales numbers also tend to increase. There is a positive correlation between advertising spending and sales.*

---

Here are some prompts to get you started with visualizations. Feel free to add additional visualizations as you see fit:
1. Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features.
    - Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).
    - Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).
2. Visualize distributions using histograms. If you have a lot, consider writing a custom function and use subplots.
    - *OPTIONAL*: Summarize the underlying distributions of your features (in words & statistics)
         - Be thorough in your verbal description of these distributions.
         - Be sure to back up these summaries with statistics.
         - We generally assume that data we sample from a population will be normally distributed. Do we observe this trend? Explain your answers for each distribution and how you think this will affect estimates made from these data.
3. Plot and interpret boxplots. 
    - Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.
    - Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).
    - Each boxplot should:
        - Only include variables of a similar scale
        - Have clear labels for each variable
        - Have appropriate titles and labels
4. Plot and interpret scatter plots to view relationships between features. Feel free to write a custom function, and subplot if you'd like. Functions save both time and space.
    - Your plots should have:
        - Two clearly labeled axes
        - A proper title
        - Colors and symbols that are clear and unmistakable
5. Additional plots of your choosing.
    - Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

In [ ]:
# Code

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Make sure to answer your question of interest or address your problem statement here.

**To-Do:** *Edit this cell with your conclusions and recommendations.*

**Don't forget to create your README!**

**To-Do:** *If you combine your problem statement, data dictionary, brief summary of your analysis, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.* Don't forget to cite your data sources!